### PR TITLE
[IMM32] Rewrite ImmRegisterWordW

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Rewrite `ImmRegisterWordW` function.

### CODE REVIEW

File: dll/win32/imm32/imm.c [hunk 1]
```diff
@@ -3411,28 +3411,56 @@ BOOL WINAPI ImmRegisterWordA(
 BOOL WINAPI ImmRegisterWordW(
   HKL hKL, LPCWSTR lpszReading, DWORD dwStyle, LPCWSTR lpszRegister)
 {
-    ImmHkl *immHkl = IMM_GetImmHkl(hKL);
-    TRACE("(%p, %s, %d, %s):\n", hKL, debugstr_w(lpszReading), dwStyle,
-                    debugstr_w(lpszRegister));
-    if (immHkl->hIME && immHkl->pImeRegisterWord)
+    PIMEDPI pImeDpi;
+    LPSTR pszReadingA = NULL, pszRegisterA = NULL;
+    BOOL ret = FALSE;
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3416]:
```suggestion
    BOOL ret = FALSE;
    PIMEDPI pImeDpi;
    LPSTR pszReadingA = NULL, pszRegisterA = NULL;
```

File: dll/win32/imm32/imm.c [hunk 2]
```diff
@@ -3411,28 +3411,56 @@ BOOL WINAPI ImmRegisterWordA(
...
+        cchW = lstrlenW(lpszRegister);
+        cchA = (cchW + 1) * 2;
+        pszRegisterA = Imm32HeapAlloc(0, cchA);
+        if (!pszRegisterA)
+            goto Quit;
+        cchA = WideCharToMultiByte(CP_ACP, 0, lpszRegister, cchW, pszRegisterA, cchA, NULL, NULL);
+        pszRegisterA[cchA] = 0;
+    }
+
+    ret = ImmRegisterWordA(hKL, pszReadingA, dwStyle, pszRegisterA);
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3455]: I see from the code that `ImmRegisterWordA` calls `ImmRegisterWordW`, but then here you call `ImmRegisterWordA` from `ImmRegisterWordW` : isn't that a problem??

File: dll/win32/imm32/imm.c [hunk 3]
```diff
@@ -3411,28 +3411,56 @@ BOOL WINAPI ImmRegisterWordA(
...
+        cchA = WideCharToMultiByte(CP_ACP, 0, lpszReading, cchW, pszReadingA, cchA, NULL, NULL);
+        pszReadingA[cchA] = 0;
     }
-    else
-        return FALSE;
+
+    if (lpszRegister)
+    {
+        cchW = lstrlenW(lpszRegister);
+        cchA = (cchW + 1) * 2;
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3447]:
```suggestion
        cchA = (cchW + 1) * sizeof(WCHAR);
```

File: dll/win32/imm32/imm.c [hunk 4]
```diff
@@ -3411,28 +3411,56 @@ BOOL WINAPI ImmRegisterWordA(
...
-            rc = immHkl->pImeRegisterWord((LPCWSTR)lpszaReading,dwStyle,
-                                          (LPCWSTR)lpszaRegister);
-            HeapFree(GetProcessHeap(),0,lpszaReading);
-            HeapFree(GetProcessHeap(),0,lpszaRegister);
-            return rc;
-        }
+    if (lpszReading)
+    {
+        cchW = lstrlenW(lpszReading);
+        cchA = (cchW + 1) * 2;
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3436]:
```suggestion
        cchA = (cchW + 1) * sizeof(WCHAR);
```

File: dll/win32/imm32/imm.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK.

File: dll/win32/imm32/imm.c (hunk 3 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK.

File: dll/win32/imm32/imm.c (hunk 4 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK.

File: dll/win32/imm32/imm.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c:3455]: Those processing depends on `fdwProperty & IME_PROP_UNICODE`. So it is not infinite loop.

Comment (katahiromz) [reply dll/win32/imm32/imm.c]: Improved.

**Pull request merged**